08-Interop: Add task to exchange UDP packets with a Zephyr node (#149)
RIOT-OS/Release-Specs: 08-interop/08-interop.md

Link-local UDP over IPv6 packet exchange (payload length 8) between an iotlab-m3
node running RIOT with lwIP and an iotlab-m3 node running RIOT with emb6 (in
both directions).

Task #11 - UDP exchange between iotlab-m3 and Zephyr
=====================================================
### Description

UDP packet exchange between an iotlab-m3 node running RIOT and a samr21-xpro
node running Zephyr.

### Testing procedure

#### Setting up the Zephyr node

1. Follow the
['Getting Started'](https://docs.zephyrproject.org/latest/getting_started/index.html)
guide on Zephyr's documentation.

> 

Reviewer: Newline not needed (in fact it unnecessarily spreads out the numbered list.

Author: Removed newlines

   Multicast Listener Discovery support      : enabled
   Neighbor cache support                    : enabled
   Neighbor discovery support                : enabled
   Duplicate address detection (DAD) support : enabled
   Router advertisement RDNSS option support : enabled
   6lo header compression support            : enabled
   Max number of IPv6 network interfaces in the system          : 1
   Max number of unicast IPv6 addresses per network interface   : 3
   Max number of multicast IPv6 addresses per network interface : 4
   Max number of IPv6 prefixes per network interface            : 2

   IPv6 addresses for interface 0x20007d60 (IEEE 802.15.4)
   =======================================================
   Type            State           Lifetime (sec)  Address
   autoconf        preferred       infinite        fe80::d419:100:7ae4:9b3b/128
   manual          preferred       infinite        2001:db8::1/128
   ```

   Similarly, to get the IEEE 802.15.4 PAN ID and channel, run:

   ```sh
   # On samr21-xpro:zephyr node
   uart:~$ ieee802154 get_pan_id
   PAN ID 43981 (0xabcd)

   uart:~$ ieee802154 get_chan
   Channel 26
   ```

#### Setting up the RIOT node

> 1. Flash on the iotlab-m3 board the [gnrc_networking] example or the [gnrc_udp]

Reviewer: ```suggestion
1. Flash the [gnrc_networking] example or the [gnrc_udp] test to the iotlab-m3 board
```

   test.

> 2. Set the channel and PAN ID to the same values as the Zephyr node, running:

Reviewer: ```suggestion
2. Set the channel and PAN ID to the same values as the Zephyr node:
```

   ```sh
   # On iotlab-m3:riot node
   > ifconfig 6 set chan 26
   ifconfig 6 set chan 26
   success: set channel on interface 6 to 26

   > ifconfig 6 set pan_id 0xabcd
   ifconfig 6 set pan_id 0xabcd
   success: set network identifier on interface 6 to 0xabcd
   ```

> 3. The UDP server application on the Zephyr node will be listening to port 4242.

Reviewer: ```suggestion
3. The UDP server application on the Zephyr node will be listening on port 4242.
```

>    Start the UDP server on the same port on the RIOT node:

Reviewer: Does the echo server reply to that port?

Author: Yes, it echoes to the same port it is listening on.

Reviewer: That's weird... why not to the source port?

Author: Sorry, I confused myself with another example I was reading. I checked [the code](https://github.com/zephyrproject-rtos/zephyr/blob/ba0aea3cd3f4aee1f21d3e1b798e6713827465f3/samples/net/sockets/echo_server/src/udp.c#L118-L119) and it sends back to the address as is, so it should be the source port.

Reviewer: Ah, but we use the same source port as the destination port for `udp send`, right! Still, something like this should be mentioned here in my opinion to avoid confusion.

Author: Changed to clarify this point.

   ```sh
   # On iotlab-m3:riot node
   > udp server start 4242
   udp server start 4242
   Success: started UDP server on port 4242
   ```

4. Send packets to the echo server in the Zephyr node, they should be echoed and
   printed on the RIOT side. Zephyr node will inform of the received packet as
   well on the shell.

   ```sh
   # On iotlab-m3:riot node

>    > udp send fe80::d419:100:7ae4:9b3b 4242 "RIOT Testing!"

Reviewer: Maybe provide the expected output here.

Author: Added output

   Multicast Listener Discovery support      : enabled
   Neighbor cache support                    : enabled
   Neighbor discovery support                : enabled
   Duplicate address detection (DAD) support : enabled
   Router advertisement RDNSS option support : enabled
   6lo header compression support            : enabled
   Max number of IPv6 network interfaces in the system          : 1
   Max number of unicast IPv6 addresses per network interface   : 3
   Max number of multicast IPv6 addresses per network interface : 4
   Max number of IPv6 prefixes per network interface            : 2

   IPv6 addresses for interface 0x20007d60 (IEEE 802.15.4)
   =======================================================
   Type            State           Lifetime (sec)  Address
   autoconf        preferred       infinite        fe80::d419:100:7ae4:9b3b/128
   manual          preferred       infinite        2001:db8::1/128
   ```

   Similarly, to get the IEEE 802.15.4 PAN ID and channel, run:

   ```sh
   # On samr21-xpro:zephyr node
   uart:~$ ieee802154 get_pan_id
   PAN ID 43981 (0xabcd)

   uart:~$ ieee802154 get_chan
   Channel 26
   ```

#### Setting up the RIOT node

> 1. Flash the [gnrc_networking] example or the [gnrc_udp] test to the iotlab-m3 board

Reviewer 2: Why not be specific about the test application?

Reviewer: Both _should_ be the same basically. The only difference is that `gnrc_udp` expects a number for the payload (representing the payload length) and `gnrc_networking` a string (representing the actual payload).

Reviewer 2: If it should be the same, then I don't see the difference in selecting one of both. If there is a relevant  difference, then we should consider testing both.

Reviewer: > If it should be the same, then I don't see the difference in selecting one of both. If there is a relevant difference, then we should consider testing both.

The only reason `gnrc_udp` exists in the first place is because the release specs were asking for large payloads that were not doable with a string input and 64 byte shell buffer ;-). So when it comes to release specs one should prefer `gnrc_udp`.

   test.
2. Set the channel and PAN ID to the same values as the Zephyr node:
   ```sh
   # On iotlab-m3:riot node

>    > ifconfig 6 set chan 26

Reviewer 2: With *gnrc_networking* the interface number is 7.

Reviewer: This should be solved once https://github.com/RIOT-OS/RIOT/pull/12738 and we adapt the `ifconfig` shell command for it (which due to the test scripts might be a quite massive change). :sweat_smile:

   ifconfig 6 set chan 26
   success: set channel on interface 6 to 26

   > ifconfig 6 set pan_id 0xabcd
   ifconfig 6 set pan_id 0xabcd
   success: set network identifier on interface 6 to 0xabcd
   ```
3. The UDP echo server application on the Zephyr node will be listening on port
   4242, and echoing to the source port of the incoming packet. As both
   [gnrc_networking] example and [gnrc_udp] test applications use the
   destination port as the source port for the `udp send` command, we need to
   start a server on port 4242:

   ```sh
   # On iotlab-m3:riot node
   > udp server start 4242
   udp server start 4242
   Success: started UDP server on port 4242
   ```
4. Send packets to the echo server in the Zephyr node, they should be echoed and
   printed on the RIOT side. Zephyr node will inform of the received packet as
   well on the shell.

   ```sh
   # On iotlab-m3:riot node

>    > udp send fe80::d419:100:7ae4:9b3b 4242 "RIOT Testing!"

Reviewer 2: This address may change so shouldn't we rather put something like ´<IPv6 addr>` here?

Reviewer: We use example addresses all the time... But maybe to make it more clearer this indeed should be `<target link-local address>` or something like that.
